lib/controllers/issueish-detail-controller.js

```diff
@@ -56,44 +58,65 @@ export class BareIssueishDetailController extends React.Component {
     this.updateTitle();
   }
 
+  componentWillMount() {
+    const {repository} = this.props;
+    if (repository && repository.pullRequest && repository.pullRequest.__typename) {
+      this.setState({typename: repository.pullRequest.__typename});
+    }
```

Comment:
Do we ever set `{typename}` to "issue"? From the diff I'm seeing, it looks like it'll either be `'PullRequest'` or `null`.

That'll work because all of the `if(typename)` clauses compare to `'PullRequest'` and have the issue case in the `else` block. It might be confusing if you saw that in the React dev tools, though.

Reply:
soooo, I think that we fetch both the `Issue` and the `PullRequest` every time, and the `PullRequest` has a `__typename` of `Issue` if it's an issue?  I should verify that, though. 

What I really wanted was to pull `__typename` up a level, so that it belongs to `repository`, but I couldn't figure out how to express that in Relay.

Reply:
just verified that my initial comment is correct -- if we're rendering an issue, we have a `pullRequest` with the `__typename` of `Issue`, and also an `issue` with the `__typename` of `Issue`.

I can see that potentially being confusing or unexpected, though.  I'll try again to pull the `__typename` up a level in the graphql so it's a property of the `repository`.

Reply:
Ohhh. Oh, I see now.

> I'll try again to pull the `__typename` up a level in the graphql so it's a property of the repository.

Hmm I don't know if you can do that or not. Maybe we could just add a comment here?

Reply:
I think we can actually get away with not using state entirely. According to [best practices](https://github.com/uberVU/react-guide/blob/master/props-vs-state.md#does-this-go-inside-props-or-state), we want to reserve our usage of state for attributes that the component will mutate at some point. For attributes that are received from above and are immutable we want to use props. In this case the typename will never be modified by the component.

I think this also increases readability and precision in our logic...

```
    if (this.state.typename !== 'PullRequest') {
      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'Cannot check out an issue');
    }
```

would instead be 

```
    if (issueish.__typename === 'Issue') {
      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'Cannot check out an issue');
    }
```

Additionally, there is some tweaking to our graphql fragment that will make this logic cleaner. I'll comment on that separately.